pkg/lib/crd/version.go

```diff
@@ -0,0 +1,52 @@
+package crd
+
+import (
+	"fmt"
+	"reflect"
+	"strings"
+
+	v1 "k8s.io/apiextensions-apiserver/pkg/apis/apiextensions/v1"
+	"k8s.io/apimachinery/pkg/apis/meta/v1/unstructured"
+	"k8s.io/apimachinery/pkg/util/yaml"
+)
+
+// V1Beta1 refers to the deprecated v1beta version of CRD objects
+const V1Beta1Version string = "v1beta1"
+
+// V1 refers to the new v1 version of CRD objects
+const V1Version string = "v1"
+
+// Version takes a CRD manifest and determines whether it is v1 or v1beta1 type.
+func Version(manifest *string) (string, error) {
+	dec := yaml.NewYAMLOrJSONDecoder(strings.NewReader(*manifest), 10)
+	unst := &unstructured.Unstructured{}
+	if err := dec.Decode(unst); err != nil {
+		return "", err
+	}
+
+	if unst.GetObjectKind().GroupVersionKind().Version == V1Beta1Version {
+		return V1Beta1Version, nil
+	}
+
+	if unst.GetObjectKind().GroupVersionKind().Version == V1Version {
+		return V1Version, nil
+	}
+
+	return "", fmt.Errorf("could not determine CRD version from manifest")
+}
+
+// NotEqual determines whether two CRDs are equal based on the versions and validations of both.
+// If false, then we know we need to update the CRD on cluster.
+func NotEqual(currentCRD *v1.CustomResourceDefinition, crd *v1.CustomResourceDefinition) bool {
+	var equalVersions bool
+	var equalValidation bool
+
+	equalVersions = reflect.DeepEqual(currentCRD.Spec.Versions, crd.Spec.Versions)
+
+	for i := range currentCRD.Spec.Versions {
+		validation := currentCRD.Spec.Versions[i].Schema
+		equalValidation = reflect.DeepEqual(validation, crd.Spec.Versions[i].Schema)
```

Comment:
could this panic of the length of the currentCRD validation list if longer than that of the previous CRD?